### PR TITLE
Gflow from pattern

### PR DESCRIPTION
**Pull request by d1ssk**
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `tox`)
- format added code by `black -l 120 <filename>`
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).

Then, please fill in below:

**Context (if applicable):**
Current gflow visualization method is based on gflow obtained from `graphix.gflow.gflow`, which is not always consistent with the pattern. #109

**Description of the change:**
Add a method to obtain gflow which is consistent with the feedforward structure of the patterns. Integrate this to the visualization method which shows gflow structure based on the pattern with the option `pattern.draw_graph(gflow_from_pattern=True)`
Also, add an option `meas_plane_indicator` to the visualization. If True, measurement plane for each node is showed adjacent to the node.

**Related issue:**
#109

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

### CODE REVIEW

**Comment by shinich1** [comment]
Nice, would you be happy to add type hints to `visualization.py`?

File: graphix/gflow.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 from graphix.linalg import MatGF2
 
 
-def gflow(graph, input, output, meas_planes, mode="single"):
+def gflow(graph, input, output, meas_planes, mode="single", pattern=None):
```
Comment (shinich1) [inline graphix/gflow.py:22]: I don't think it's a good idea to have pattern as optional argument in `gflow`. having `gflow_from_pattern` is enough. can we get rid of added codes to `gflow`?

File: graphix/gflow.py [hunk 2]
```diff
@@ -207,6 +214,55 @@ def gflowaux(
...
+                    g[g_out] = {g_in}
+        if cmd[0] == "X":
+            g_in = cmd[1]
+            g_outs = set(cmd[2]) & nodes
+            for g_out in g_outs:
+                if g_out in g.keys():
+                    g[g_out] = g[g_out].union({g_in})
+                else:
+                    g[g_out] = {g_in}
+
```
Comment (shinich1) [inline graphix/gflow.py:262]: Could you add gflow check (based on #112) at the end of this function?

File: graphix/pattern.py [hunk 3]
```diff
@@ -1234,9 +1234,11 @@ def draw_graph(
         self,
         node_distance=(1, 1),
         figsize=None,
+        gflow_from_pattern=False,
```
Comment (shinich1) [inline graphix/pattern.py:1237]: perhaps `gflow_from_pattern` should be defaulted to True

File: graphix/gflow.py (hunk 1 above)
Comment (d1ssk) [reply graphix/gflow.py]: This was difinitely unnecessary. I deleted it. Thank you.

File: graphix/pattern.py (hunk 3 above)
Comment (d1ssk) [reply graphix/pattern.py]: I set the default to True.

**Comment by shinich1** [comment]
@d1ssk let's merge this once the additional error reported in #107 is fixed, as #112 may take some time to complete. we'll add the gflow check later.

**Comment by d1ssk** [comment]
Understood. I have already corrected the additional error and I'm ready for merging this after your final review.

File: graphix/pattern.py [hunk 4]
```diff
@@ -1234,9 +1234,11 @@ def draw_graph(
         self,
         node_distance=(1, 1),
         figsize=None,
+        gflow_from_pattern=True,
         pauli_indicator=True,
         show_loop=True,
         local_clifford_indicator=False,
+        meas_plane_indicator=False,
```
Comment (shinich1) [inline graphix/pattern.py:1241]: change to show_measurement_planes

File: graphix/visualization.py [hunk 5]
```diff
@@ -344,7 +409,15 @@ def visualize_w_gflow(
             plt.savefig(filename)
         plt.show()
 
-    def visualize_wo_structure(self, angles=None, local_clifford=None, node_distance=(1, 1), save=False, filename=None):
+    def visualize_wo_structure(
```
Comment (shinich1) [inline graphix/visualization.py:412]: Is this currently showing entirely by random node placements? Is it possible to at least arrange nodes by their preparation or measurement order, perhaps (no need to be very tidy as there's no structure in this case

**Review by shinich1 — APPROVED** [review]
Good work! Let's merge once the additional comments are resolved.

File: graphix/visualization.py (hunk 5 above)
Comment (d1ssk) [reply graphix/visualization.py:412]: I will address this in a separate issue.

**Comment by shinich1** [comment]
@d1ssk lets squash and merge.

**Pull request merged**